File: mobile/test/data/repositories/server_discovery_repository_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:wheeldeck/data/repositories/server_discovery_repository.dart';
import 'package:wheeldeck/domain/models/connection_mode.dart';
import 'package:wheeldeck/network/discovery.dart';

void main() {
  group('ServerDiscoveryRepository', () {
    test('refresh caches servers and exposes immutable snapshots', () async {
      final repository = ServerDiscoveryRepository(
        discovery: ServerDiscovery(
          resolve: () async => const [
            DiscoveredServer(host: '10.0.0.1', port: 8765, name: 'Desktop'),
          ],
        ),
      );

      expect(repository.servers, isEmpty);

      final servers = await repository.refresh();

      expect(servers, hasLength(1));
      expect(repository.servers.single.name, 'Desktop');
      expect(
        () => (repository.servers as List).add(
          const DiscoveredServer(host: 'x', port: 1, name: 'y'),
        ),
        throwsUnsupportedError,
      );
    });

    test('manualTarget builds a manual connection target', () {
      final target = ServerDiscoveryRepository.manualTarget(
        host: '192.168.1.7',
        port: 9000,
      );

      expect(target.mode, ConnectionMode.manual);
      expect(target.ipAddress, '192.168.1.7');
      expect(target.port, 9000);
    });
  });
}

File: android/app/src/test/kotlin/dev/fazrigading/wheeldeck/data/repositories/ServerDiscoveryRepositoryTest.kt
package dev.fazrigading.wheeldeck.data.repositories

import dev.fazrigading.wheeldeck.data.services.ServerDiscovery
import dev.fazrigading.wheeldeck.domain.models.ConnectionMode
import dev.fazrigading.wheeldeck.domain.models.DiscoveredServer
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Test

class ServerDiscoveryRepositoryTest {

    @Test
    fun `refresh caches the sweep and exposes a snapshot`() = runTest {
        val first = listOf(DiscoveredServer(host = "192.168.1.10", port = 8765, name = "Desktop"))
        var results = first
        val repository = ServerDiscoveryRepository(ServerDiscovery { results })

        assertEquals(first, repository.refresh())
        assertEquals(first, repository.servers)

        val second = listOf(
            DiscoveredServer(host = "192.168.1.10", port = 8765, name = "Desktop"),
            DiscoveredServer(host = "192.168.1.20", port = 8765, name = "Laptop"),
        )
        results = second
        assertEquals(second, repository.refresh())
        assertEquals(second, repository.servers)
    }

    @Test
    fun `servers snapshot is isolated from later refreshes`() = runTest {
        val first = listOf(DiscoveredServer(host = "192.168.1.10", port = 8765, name = "Desktop"))
        var results = first
        val repository = ServerDiscoveryRepository(ServerDiscovery { results })
        repository.refresh()
        val snapshot = repository.servers

        results = emptyList()
        repository.refresh()

        assertEquals(first, snapshot)
        assertEquals(emptyList<DiscoveredServer>(), repository.servers)
    }

    @Test
    fun `manual target delegates to the service`() {
        val target = ServerDiscoveryRepository.manualTarget(host = "192.168.1.50", port = 9000)

        assertEquals(ConnectionMode.Manual, target.mode)
        assertEquals(9000, target.port)
    }
}
